New package for `yubikey-manager-qt`

avdv opened the pull request:
###### Motivation for this change

I own a few Yubikeys, USB keys for FIDO2 U2F auth. There is a new graphical  manager appliation based on Qt5 (similar to yubikey-manager-neo) from Yubico.

This is my first try writing a Nix expression for a new package, a probably made some mistakes. Please advice.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

avdv committed:
maintainers: add cbley

avdv committed:
yubikey-manager-qt: init at 1.1.0

delroth commented on pkgs/tools/misc/yubikey-manager-qt/default.nix:
```diff
@@ -0,0 +1,81 @@
+, yubikey-personalization
+}:
+
+with import <nixpkgs> {};
```
Not needed within nixpkgs.

avdv replied:
OK, removed.

delroth commented on pkgs/tools/misc/yubikey-manager-qt/default.nix:
```diff
@@ -0,0 +1,81 @@
+  version = "1.1.0";
+
+  src = fetchurl {
+    url = "https://developers.yubico.com/yubikey-manager-qt/Releases/yubikey-manager-qt-${version}.tar.gz";
```
Just ${name}.tar.gz?

avdv replied:
Yes.

delroth commented on pkgs/tools/misc/yubikey-manager-qt/default.nix:
```diff
@@ -0,0 +1,81 @@
+    substituteInPlace ykman-gui/deployment.pri --replace '/usr/bin' "$out/bin"
+  '';
+
+  buildInputs = [ pythonPackages.python stdenv qt5.qtbase qt5.qtgraphicaleffects qt5.qtquickcontrols qt5.qtquickcontrols2 pyotherside ];
```
Pretty sure you don't need to explicitly list stdenv is there.

avdv replied:
Removed.

delroth commented on pkgs/tools/misc/yubikey-manager-qt/default.nix:
```diff
@@ -0,0 +1,81 @@
+    license = licenses.bsd2;
+    maintainers = [ maintainers.cbley ];
+    platforms = platforms.linux;
+    homepage = "https://github.com/Yubico";
```
Duplicate attribute.

avdv replied:
Oh, that was silly, wasn't it!? I should not change something without running `nix-build` again...

Ma27 commented on pkgs/tools/misc/yubikey-manager-qt/default.nix:
```diff
@@ -0,0 +1,81 @@
+  ];
+
+in stdenv.mkDerivation rec {
+  name = "${pname}-${version}";
```
This is actually not needed if `pname` and `version` is set.

Ma27 commented on pkgs/tools/misc/yubikey-manager-qt/default.nix:
```diff
@@ -0,0 +1,81 @@
+    license = licenses.bsd2;
+    maintainers = [ maintainers.cbley ];
+    platforms = platforms.linux;
+    homepage = "https://github.com/Yubico";
```
duplicate keys in an attribute set always break evaluation.

Ma27 commented on pkgs/tools/misc/yubikey-manager-qt/default.nix:
```diff
@@ -0,0 +1,81 @@
+
+  buildPhase = ''
+    qmake
+    make
```
Manually invoking `qmake` and `make` shouldn't be needed.

avdv replied:
Oh, of course. That's just because I started from a "normal" `callPackage` until I found out about the `libsForQt5.callPackage` helper.

delroth commented on pkgs/tools/misc/yubikey-manager-qt/default.nix:
```diff
@@ -0,0 +1,81 @@
+  ];
+
+in stdenv.mkDerivation rec {
+  name = "${pname}-${version}";
```
Doesn't seem like pname is used anywhere else, just inline it in here?

avdv replied:
That seems to be a standard attribute, given what @Ma27 said about the `meta.name` attribute.

Ma27 replied:
I obviously didn't explain this well enough: you can either use `pname` (and `version`) which is concatenated to `${pname}-${version}` or simply use a `name` attribute. The first option has been implemented for convenience reasons, the latter is more commonly used.

Ma27 commented on pkgs/tools/misc/yubikey-manager-qt/default.nix:
```diff
@@ -0,0 +1,81 @@
+    wrapProgram $out/bin/ykman-gui \
+      --prefix PYTHONPATH : "$program_PYTHONPATH" \
+      --prefix LD_PRELOAD : "${yubikey-personalization}/lib/libykpers-1.so" \
+      --prefix LD_LIBRARY_PATH : "${stdenv.lib.getLib pcsclite}/lib:${yubikey-personalization}/lib" \
```
Actually an appropriate ELF header is produced during the build. But I need to recheck this as soon as we got this working properly :)

Ma27 replied:
Again, isn't `pcsclite` a dependency (and shouldn't be added using `LD_LIBRARY_PATH`)? Also I'm not sure if `LD_PRELOAD` is actually needed.

delroth commented on pkgs/tools/misc/yubikey-manager-qt/default.nix:
```diff
@@ -0,0 +1,81 @@
+
+  meta = with stdenv.lib; {
+    inherit version;
+    description = ''Cross-platform application for configuring any YubiKey over all USB interfaces.'';
```
description = ''
      Cross-platform application for configuring any YubiKey over all USB interfaces.
    '';

Or

    description = "Cross-platform application for configuring any YubiKey over all USB interfaces.";

delroth reviewed (commented):
Did a quick first review pass -- I don't know anything about Qt or Python packaging in Nix, so mostly addressing the low hanging fruits :-)

Ma27 reviewed (changes requested):
Thanks a lot for your contribution. Unfortunately I get the following error when starting `./result/bin/ykman-gui` (after removing the duplicated `homepage` entry from `meta`):

```
QQmlApplicationEngine failed to load component
qrc:/qml/main.qml:1 module "QtQuick" is not installed
qrc:/qml/main.qml:6 module "QtQuick.Controls.Material" is not installed
qrc:/qml/main.qml:4 module "Qt.labs.settings" is not installed
qrc:/qml/main.qml:2 module "QtQuick.Controls" is not installed
qrc:/qml/main.qml:7 module "QtQuick.Window" is not installed
qrc:/qml/main.qml:3 module "QtQuick.Layouts" is not installed
qrc:/qml/main.qml:5 module "Qt.labs.platform" is not installed
qrc:/qml/main.qml:1 module "QtQuick" is not installed
qrc:/qml/main.qml:6 module "QtQuick.Controls.Material" is not installed
qrc:/qml/main.qml:4 module "Qt.labs.settings" is not installed
qrc:/qml/main.qml:2 module "QtQuick.Controls" is not installed
qrc:/qml/main.qml:7 module "QtQuick.Window" is not installed
qrc:/qml/main.qml:3 module "QtQuick.Layouts" is not installed
qrc:/qml/main.qml:5 module "Qt.labs.platform" is not installed
qrc:/qml/main.qml:1 module "QtQuick" is not installed
qrc:/qml/main.qml:6 module "QtQuick.Controls.Material" is not installed
qrc:/qml/main.qml:4 module "Qt.labs.settings" is not installed
qrc:/qml/main.qml:2 module "QtQuick.Controls" is not installed
qrc:/qml/main.qml:7 module "QtQuick.Window" is not installed
qrc:/qml/main.qml:3 module "QtQuick.Layouts" is not installed
qrc:/qml/main.qml:5 module "Qt.labs.platform" is not installed
qrc:/qml/main.qml:1 module "QtQuick" is not installed
qrc:/qml/main.qml:6 module "QtQuick.Controls.Material" is not installed
qrc:/qml/main.qml:4 module "Qt.labs.settings" is not installed
qrc:/qml/main.qml:2 module "QtQuick.Controls" is not installed
qrc:/qml/main.qml:7 module "QtQuick.Window" is not installed
qrc:/qml/main.qml:3 module "QtQuick.Layouts" is not installed
qrc:/qml/main.qml:5 module "Qt.labs.platform" is not installed
qrc:/qml/main.qml:1 module "QtQuick" is not installed
qrc:/qml/main.qml:6 module "QtQuick.Controls.Material" is not installed
qrc:/qml/main.qml:4 module "Qt.labs.settings" is not installed
qrc:/qml/main.qml:2 module "QtQuick.Controls" is not installed
qrc:/qml/main.qml:7 module "QtQuick.Window" is not installed
qrc:/qml/main.qml:3 module "QtQuick.Layouts" is not installed
qrc:/qml/main.qml:5 module "Qt.labs.platform" is not installed
qrc:/qml/main.qml:1 module "QtQuick" is not installed
qrc:/qml/main.qml:6 module "QtQuick.Controls.Material" is not installed
qrc:/qml/main.qml:4 module "Qt.labs.settings" is not installed
qrc:/qml/main.qml:2 module "QtQuick.Controls" is not installed
qrc:/qml/main.qml:7 module "QtQuick.Window" is not installed
qrc:/qml/main.qml:3 module "QtQuick.Layouts" is not installed
qrc:/qml/main.qml:5 module "Qt.labs.platform" is not installed
qrc:/qml/main.qml:1 module "QtQuick" is not installed
qrc:/qml/main.qml:6 module "QtQuick.Controls.Material" is not installed
qrc:/qml/main.qml:4 module "Qt.labs.settings" is not installed
qrc:/qml/main.qml:2 module "QtQuick.Controls" is not installed
qrc:/qml/main.qml:7 module "QtQuick.Window" is not installed
qrc:/qml/main.qml:3 module "QtQuick.Layouts" is not installed
qrc:/qml/main.qml:5 module "Qt.labs.platform" is not installed

[1]    24343 segmentation fault  ./result/bin/ykman-gui
```

Ma27 commented:
...and @delroth was slightly faster than me :sweat_smile:

avdv committed:
fixup! yubikey-manager-qt: init at 1.1.0

avdv commented:
@Ma27 and @delroth thank you so much! I (hopefully) fixed all the issues you mentioned and pushed a fixup commit. I'll squash the last two commits on your mark.

Ma27 commented:
Thanks a lot for stepping through the feedback. Although the program starts for me, it doesn't detect a YubiKey for me (`yubikey-personalization-gui` is installed in my `nixos-config` and detects it). Am I missing any additional `udev` rules? In that case we should document this i.e. in `doc/package-notes.xml`.

avdv commented:
> Thanks a lot for stepping through the feedback. Although the program starts for me, it doesn't detect a YubiKey for me (`yubikey-personalization-gui` is installed in my `nixos-config` and detects it). Am I missing any additional `udev` rules? In that case we should document this i.e. in `doc/package-notes.xml`.

I also had a hard time to figure this out...

You need to have the pcscd service (from pcsclite) configured for the program to work. (`services.systemd.pcslite.enabled = true;` if I remember correctly -- I am not at home right now)

And you need the udev rules from `libu2f-host` enabled for Firefox to recognize the stick for authentication. (`services.udev.packages = [ libu2f-hosts ];')

Ma27 commented:
The service you're reffering to is called `pcscd`. With it enabled the GUI seems to work :+1:

dtzWill commented:
Works here too, thanks!

Ma27 committed:
yubikey-manager-qt: cleanup

* Explicitly specify all QT dependencies rather than import from the
  `qt5` attr set. This makes overrides of a single library easier.

* Drop the superfluous `with stdenv` expression and reference lib or
  stdenv itself where possible.

* Don't manually configure shared libraries to load. This is mostly done
  automatically during the build steps.

Ma27 commented:
I pushed a second commit onto your branch (that can be easily removed if needed) that does some minor cleanup as explained in the commit message. If the commit is fine for you and with it the package remains functional for you, I'd approve and merge :)

avdv commented:
@Ma27 LGTM, and still works for me, thank you!

Ma27 commented:
@avdv thanks!
